storm-client/src/jvm/org/apache/storm/stats/StatsUtil.java

```diff
@@ -1525,27 +1528,24 @@ public static ComponentPageInfo aggCompExecsStats(
      * @param timeout       timeout
      * @return a HashMap of updated executor heart beats
      */
-    public static Map<List<Integer>, Map<String, Object>> updateHeartbeatCacheFromZkHeartbeat(Map<List<Integer>, Map<String, Object>> cache,
-                                                                                              Map<List<Integer>, Map<String, Object>>
-                                                                                                  executorBeats,
-                                                                                              Set<List<Integer>> executors,
-                                                                                              Integer timeout) {
-        Map<List<Integer>, Map<String, Object>> ret = new HashMap<>();
-        if (cache == null && executorBeats == null) {
-            return ret;
-        }
-
+    public static ConcurrentMap<List<Integer>, Map<String, Object>> updateHeartbeatCacheFromZkHeartbeat(Map<List<Integer>, Map<String, Object>> cache,
+                                                                                                        Map<List<Integer>, Map<String, Object>>
+                                                                                                                executorBeats,
+                                                                                                        Set<List<Integer>> executors,
+                                                                                                        Integer timeout) {
         if (cache == null) {
+            if (executorBeats == null) {
```

Comment:
The construction here seems a little odd. Initializing cache and executorBeats when null isn't necessary. I think we should keep the "if cache and executorBeats are null" clause, then replace the other two branches with something like `Map<String, Object> currBeat = cache == null ? null : cache.get(executor);` and equivalent for `newBeat`.

Reply:
Sorry, i did see any reason why this method is not thread safe, cause it almost a tool method, only to initialize a Map cache which is updated into `Nimbus` heartbeatsCache through `heartbeatsCache.getAndUpdate(new Assoc<>(topoId, cache))`,` ConcurrentModificationException` happens when we iterate over a collection through iterator and also modify it, but here, we only iterate the executor list and do not modify any of the list entry.